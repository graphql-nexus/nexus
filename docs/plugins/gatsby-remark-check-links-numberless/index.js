var visit = require('unist-util-visit')
const path = require('path')

function getCacheKey(node) {
  return `remark-check-links-${node.id}-${node.internal.contentDigest}`
}

function getHeadingsMapKey(link, pathUrl) {
  let key = link
  const hashIndex = link.indexOf('#')
  const hasHash = hashIndex !== -1
  if (hasHash) {
    key = link.startsWith('#') ? pathUrl : link.slice(0, hashIndex)
  }

  return {
    key,
    hasHash,
    hashIndex,
  }
}

function createPathPrefixer(pathPrefix) {
  return function withPathPrefix(url) {
    var prefixed = pathPrefix + url
    return prefixed.replace(/\/\//, '/')
  }
}

module.exports = async function plugin(
  { markdownAST, markdownNode, files, getNode, cache, getCache, pathPrefix },
  { exceptions = [], ignore = [], verbose = true } = {}
) {
  const withPathPrefix = createPathPrefixer(pathPrefix)
  const pathSep = '/'
  if (!markdownNode.fields) {
    // let the file pass if it has no fields
    return markdownAST
  }

  const links = []
  const headings = []

  function visitor(node, index, parent) {
    //To convert all uppercase links to lowercase (if used by mistake)
    node.url = node.url.toLowerCase()

    if (parent.type === 'heading') {
      headings.push(parent.data.id)
      return
    }

    if (!node.url.startsWith('mailto:') && !/^https?:\/\//.test(node.url)) {
      let tranformedUrl = node.url
      links.push({
        ...node,
        tranformedUrl,
        frontmatter: markdownNode.frontmatter,
      })
    }
  }

  visit(markdownAST, 'link', visitor)

  const parent = await getNode(markdownNode.parent)
  const setAt = Date.now()
  cache.set(getCacheKey(parent), {
    path: withPathPrefix(
      markdownNode.fields.slug
        .replace(/\/index$/, '')
        .replace(/\d+-/g, '')
        .concat(pathSep)
    ),
    links,
    headings,
    setAt,
  })

  // wait to see if all of the Markdown and MDX has been visited
  const linksMap = {}
  const headingsMap = {}
  for (const file of files) {
    if (/^mdx?$/.test(file.extension) && file.relativePath !== 'docs/README.md') {
      const key = getCacheKey(file)

      let visited = await cache.get(key)
      if (!visited && getCache) {
        // the cache provided to `gatsby-mdx` has its own namespace, and it
        // doesn't have access to `getCache`, so we have to check to see if
        // those files have been visited here.
        const mdxCache = getCache('gatsby-plugin-mdx')
        visited = await mdxCache.get(key)
      }
      if (visited && setAt >= visited.setAt) {
        linksMap[visited.path] = visited.links
        headingsMap[visited.path] = visited.headings
        continue
      }

      // don't continue if a page hasn't been visited yet
      return
    }
  }

  let totalBrokenLinks = 0
  const prefixedIgnore = ignore.map(withPathPrefix)
  const prefixedExceptions = exceptions.map(withPathPrefix)
  const pathKeys = Object.keys(linksMap)
  const pathKeysWithoutIndex = pathKeys.map(p =>
    p.replace(`${pathSep}index`, '').replace(/\/$/, '')
  )
  for (const pathL in linksMap) {
    if (prefixedIgnore.includes(pathL)) {
      // don't count broken links for ignored pages
      continue
    }

    const linksForPath = linksMap[pathL]
    if (linksForPath.length) {
      const brokenLinks = linksForPath.filter(link => {
        // return true for broken links, false = pass
        const { key, hasHash, hashIndex } = getHeadingsMapKey(link.tranformedUrl, pathL)
        if (prefixedExceptions.includes(key)) {
          return false
        }

        const url = hasHash ? link.tranformedUrl.slice(0, hashIndex) : link.tranformedUrl
        const urlToCheck = url.slice(-1) === pathSep ? url.slice(0, -1) : url
        const headings = headingsMap[key]

        if (headings) {
          if (hasHash) {
            const id = link.tranformedUrl.slice(hashIndex + 1)
            return !prefixedExceptions.includes(id) && !headings.includes(id)
          }

          return false
        }
        return !pathKeysWithoutIndex.includes(urlToCheck)
      })

      const brokenLinkCount = brokenLinks.length
      totalBrokenLinks += brokenLinkCount
      if (brokenLinkCount && verbose) {
        console.warn(`${brokenLinkCount} broken links found on ${pathL}`)
        for (const link of brokenLinks) {
          let prefix = '-'
          if (link.position) {
            const { line, column } = link.position.start

            // account for the offset that frontmatter adds
            const offset = link.frontmatter ? Object.keys(link.frontmatter).length + 2 : 0

            prefix = [String(line + offset).padStart(3, ' '), String(column).padEnd(4, ' ')].join(
              ':'
            )
          }
          console.warn(`${prefix} ${link.originalUrl}`)
        }
        console.log('')
      }
    }
  }

  if (totalBrokenLinks) {
    const message = `${totalBrokenLinks} broken links found`
    if (process.env.NODE_ENV === 'production') {
      // break builds with broken links before they get deployed for reals
      // throw new Error(message);
      console.info('Broken links found. Please fix before deploy!')
    }

    if (verbose) {
      console.error(message)
    }
  } else if (verbose) {
    console.info('No broken links found')
  }

  return markdownAST
}
