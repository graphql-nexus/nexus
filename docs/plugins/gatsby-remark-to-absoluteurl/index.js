var visit = require('unist-util-visit')
const path = require('path')

function withPathPrefix(url, pathPrefix) {
  const prefixed = pathPrefix + url
  return prefixed.replace(/\/\//, '/')
}

const pathSep = '/'

module.exports = function plugin(
  { markdownAST, markdownNode, pathPrefix, getNode },
  { redirects = [] } = {}
) {
  function visitor(node) {
    //To convert all uppercase links to lowercase (if used by mistake)
    node.url = node.url.toLowerCase()

    node.originalUrl = node.url
    if (
      markdownNode.fields &&
      markdownNode.fields.slug &&
      !node.url.startsWith('/') &&
      !node.url.startsWith('#') &&
      !node.url.startsWith('mailto:') &&
      !/^https?:\/\//.test(node.url)
    ) {
      const parent = getNode(markdownNode.parent)
      const newUrl = path
        .resolve(
          markdownNode.fields.slug
            .replace(`${pathSep}index`, '')
            .replace(/\d+-/g, '')
            .replace(/\/$/, '')
            .split(pathSep)
            .slice(0, parent.name === 'index' ? undefined : -1)
            .join(pathSep) || '/',
          node.url
        )
        .replace(/\/?(\?|#|$)/, '/$1')

      if (/^..\\/.test(newUrl)) {
        //Code specifically for local run, to fix broken links on
        let newUrl2 = path.resolve(
          markdownNode.fields.slug.replace(/(\/.+)\/.*/, '$1').replace(/\/\d+-/g, '/'),
          node.url
        )

        newUrl2 = newUrl2
          .replace(/\\/g, '/')
          .slice(2)
          .replace(/(^.*)#.*/, '$1')
        const isRedirectPath = redirects.find(url => newUrl2.includes(url.from))
        if (isRedirectPath) newUrl2 = isRedirectPath.to

        let hashVal = node.url.match(/#.*/)
        if (hashVal) newUrl2 += hashVal[0]

        node.url = newUrl2.replace(/^([^#]*)$/, '$1/')
      } else {
        const isRedirectPath = redirects.find(url => newUrl.includes(url.from))
        node.url = withPathPrefix(
          isRedirectPath ? newUrl.replace(isRedirectPath.from, isRedirectPath.to) : newUrl,
          pathPrefix
        )
      }
    }
  }

  visit(markdownAST, 'link', visitor)

  return markdownAST
}
