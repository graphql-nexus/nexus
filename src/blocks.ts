export { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
export { InterfaceDefinitionBlock } from './definitions/interfaceType'
export { ObjectDefinitionBlock } from './definitions/objectType'
export { UnionDefinitionBlock } from './definitions/unionType'
