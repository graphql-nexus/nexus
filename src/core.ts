// The "core" is used as a namespace to re-export everything,
// For anyone who wants to use the internals
export * from './builder'
export * from './makeSchema'
export * from './definitions/args'
export * from './definitions/decorateType'
export * from './definitions/definitionBlocks'
export * from './definitions/enumType'
export * from './definitions/extendInputType'
export * from './definitions/extendType'
export * from './definitions/inputObjectType'
export * from './definitions/interfaceType'
export * from './definitions/list'
export * from './definitions/mutationField'
export * from './definitions/mutationType'
export * from './definitions/nonNull'
export * from './definitions/nullable'
export * from './definitions/objectType'
export * from './definitions/queryField'
export * from './definitions/queryType'
export * from './definitions/scalarType'
export * from './definitions/subscriptionField'
export * from './definitions/subscriptionType'
export * from './definitions/nexusMeta'
export * from './definitions/unionType'
export * from './definitions/wrapping'
export * from './definitions/_types'
export * from './dynamicMethod'
export * from './plugin'
export * from './plugins'
export * from './rebuildType'
export * from './sdlConverter'
export * from './typegenAutoConfig'
export * from './typegenFormatPrettier'
export * from './typegenMetadata'
export * from './typegenPrinter'
export * from './typegenTypeHelpers'
export * from './typegenUtils'
export * from './utils'
