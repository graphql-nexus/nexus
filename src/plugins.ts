import { SchemaBuilder, NexusAcceptedTypeDef } from "./builder";
import { withNexusSymbol, NexusTypes } from "./definitions/_types";
import { venn } from "./utils";

/**
 * A read-only builder api exposed to plugins in the onInstall hook which
 * proxies very limited functionality into the internal Nexus Builder.
 */
export type PluginBuilderLens = {
  hasType: (typeName: string) => boolean;
};

/**
 * This is the Neuxs Plugin interface that allows users to extend Nexus at
 * particular extension points. Plugins are just functions that receive hooks
 * which can then be registered upon with callbacks.
 */
export type PluginConfig = {
  name: string;
  onInstall?: PluginOnInstallHandler;
};

/**
 * The plugin callback to execute when onInstall lifecycle event occurs.
 * OnInstall event occurs before type walking which means inline types are not
 * visible at this point yet. `builderLens.hasType` will only return true
 * for types the user has defined top level in their app, and any types added by
 * upstream plugins.
 */
export type PluginOnInstallHandler = (
  builder: PluginBuilderLens
) => { types: NexusAcceptedTypeDef[] };

/**
 * The processed version of a plugin config. This lower level version has
 * defaults provided for optionals etc.
 */
export type InternalPluginConfig = Required<PluginConfig>;

/**
 * A plugin defines configuration which can document additional metadata options
 * for a type definition. This metadata can be used to decorate the "resolve" function
 * to provide custom functionality, such as logging, error handling, additional type
 * validation.
 *
 * You can specify options which can be defined on the schema,
 * the type or the plugin. The config from each of these will be
 * passed in during schema construction time, and used to augment the field as necessary.
 *
 * You can either return a function, with the new defintion of a resolver implementation,
 * or you can return an "enter" / "leave" pairing which will wrap the pre-execution of the
 * resolver and the "result" of the resolver, respectively.
 */
export function createPlugin(config: PluginConfig): PluginDef {
  validatePluginConfig(config);
  const internalConfig = { ...configDefaults, ...config };
  return new PluginDef(internalConfig);
}

const configDefaults = {
  onInstall: () => ({ types: [] }),
};

/**
 * A definition for a plugin. Should be passed to the `plugins: []` option
 * on makeSchema. Refer to `createPlugin` factory for full doc.
 */
export class PluginDef {
  constructor(readonly config: InternalPluginConfig) {}
}
withNexusSymbol(PluginDef, NexusTypes.Plugin);

/**
 * The interface used to drive plugin execution via lifecycle event triggering.
 */
type PluginController = {
  triggerOnInstall: () => void;
};

/**
 * This will gather the hook handlers (aka. callbacks, event handlers) a the
 * plugin has registered for and return a controller  to trigger said hooks,
 * thus controlling execution of the plugins' hook handlers.
 */
export function initialize(
  builder: SchemaBuilder,
  plugin: PluginDef
): PluginController {
  const state = {
    onInstallTriggered: false,
  };

  const builderLens: PluginBuilderLens = {
    hasType: builder.hasType,
  };

  return {
    triggerOnInstall: () => {
      // Enforce the invariant that a lifecycle hook will only ever be called once.
      if (state.onInstallTriggered) {
        throw new Error(
          "Multiple triggers of onInstall hook detected. This should never happen. This is an internal error."
        );
      } else {
        state.onInstallTriggered = true;
      }

      // By doing addType on the types returned by a plugin right after it has
      // done so we make it possible for downstream plugins to see types added
      // by upstream plugins.
      let hookResult: ReturnType<PluginOnInstallHandler>;
      try {
        hookResult = plugin.config.onInstall(builderLens);
      } catch (error) {
        throw new Error(
          `Plugin ${plugin.config.name} failed on "onInstall" hook:\n\n${error.stack}`
        );
      }

      validateOnInstallHookResult(plugin, hookResult);
      hookResult.types.forEach(builder.addType);
    },
  };
}

/**
 * Validate that the configuration given by a plugin is valid.
 */
function validatePluginConfig(plugin: PluginConfig): void {
  const validRequiredProps = ["name"];
  const validOptionalProps = ["onInstall"];
  const validProps = [...validRequiredProps, ...validOptionalProps];
  const givenProps = Object.keys(plugin);

  const printProps = (props: Iterable<string>): string => {
    return [...props].join(", ");
  };

  const [missingRequiredProps, ,] = venn(validRequiredProps, givenProps);
  if (missingRequiredProps.size > 0) {
    throw new Error(
      `Plugin "${plugin.name}" is missing required properties: ${printProps(
        missingRequiredProps
      )}`
    );
  }

  const nameType = typeof plugin.name;
  if (nameType !== "string") {
    throw new Error(
      `Plugin "${plugin.name}" is giving an invalid value for property name: expected "string" type, got ${nameType} type`
    );
  }

  if (plugin.name === "") {
    throw new Error(
      `Plugin "${plugin.name}" is giving an invalid value for property name: empty string`
    );
  }

  const [, , invalidGivenProps] = venn(validProps, givenProps);
  if (invalidGivenProps.size > 0) {
    throw new Error(
      `Plugin "${plugin.name}" is giving unexpected properties: ${printProps(
        invalidGivenProps
      )}`
    );
  }

  if (plugin.onInstall) {
    const onInstallType = typeof plugin.onInstall;
    if (onInstallType !== "function") {
      throw new Error(
        `Plugin "${plugin.name}" is giving an invalid value for onInstall hook: expected "function" type, got ${onInstallType} type`
      );
    }
  }
}

/**
 * Validate that the data returned from a plugin from the `onInstall` hook is valid.
 */
function validateOnInstallHookResult(
  plugin: PluginDef,
  hookResult: ReturnType<PluginOnInstallHandler>
): void {
  if (
    hookResult === null ||
    typeof hookResult !== "object" ||
    !Array.isArray(hookResult.types)
  ) {
    throw new Error(
      `Plugin "${plugin.config.name}" returned invalid data for "onInstall" hook:\n\nexpected structure:\n\n  { types: NexusAcceptedTypeDef[] }\n\ngot:\n\n  ${hookResult}`
    );
  }
  // TODO we should validate that the array members all fall under NexusAcceptedTypeDef
}
