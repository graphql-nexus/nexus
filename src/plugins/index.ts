export { connectionPlugin } from './connectionPlugin'
export * as connectionPluginCore from './connectionPlugin'

export { fieldAuthorizePlugin } from './fieldAuthorizePlugin'
export * as fieldAuthorizePluginCore from './fieldAuthorizePlugin'

export { nullabilityGuardPlugin } from './nullabilityGuardPlugin'
export * as nullabilityGuardPluginCore from './nullabilityGuardPlugin'

export { queryComplexityPlugin } from './queryComplexityPlugin'
export * as queryComplexityPluginCore from './queryComplexityPlugin'

export { declarativeWrappingPlugin } from './declarativeWrappingPlugin'
