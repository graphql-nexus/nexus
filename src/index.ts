import * as blocks from './blocks'
import * as core from './core'

// All of the Public API definitions
export { makeSchema } from './makeSchema'
export { arg, booleanArg, floatArg, idArg, intArg, stringArg } from './definitions/args'
export { decorateType } from './definitions/decorateType'
export { enumType } from './definitions/enumType'
export { extendInputType } from './definitions/extendInputType'
export { extendType } from './definitions/extendType'
export { inputObjectType } from './definitions/inputObjectType'
export { interfaceType } from './definitions/interfaceType'
export { list } from './definitions/list'
export { mutationField } from './definitions/mutationField'
export { mutationType } from './definitions/mutationType'
export { nonNull } from './definitions/nonNull'
export { nullable } from './definitions/nullable'
export { objectType } from './definitions/objectType'
export { queryField } from './definitions/queryField'
export { queryType } from './definitions/queryType'
export { asNexusMethod, scalarType } from './definitions/scalarType'
export { subscriptionField } from './definitions/subscriptionField'
export { subscriptionType } from './definitions/subscriptionType'
export { unionType } from './definitions/unionType'
export { dynamicInputMethod, dynamicOutputMethod } from './dynamicMethod'
export { dynamicOutputProperty } from './dynamicProperty'
export { createPlugin, plugin, PluginBuilderLens, PluginConfig } from './plugin'
export * from './plugins'
export { convertSDL } from './sdlConverter'
export { AllInputTypes, AllOutputTypes, FieldResolver, FieldType } from './typegenTypeHelpers'
export { groupTypes } from './utils'
export { blocks, core }
