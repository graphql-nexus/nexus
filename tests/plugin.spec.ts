import { buildSchema, graphql, GraphQLSchema, lexicographicSortSchema, printSchema } from 'graphql'
import {
  interfaceType,
  list,
  makeSchema,
  MiddlewareFn,
  nonNull,
  objectType,
  plugin,
  queryField,
  booleanArg,
  inputObjectType,
} from '../src/core'
import { nullabilityGuardPlugin } from '../src/plugins'
import { EXAMPLE_SDL } from './_sdl'

const nullGuardPlugin = nullabilityGuardPlugin({
  shouldGuard: true,
  fallbackValues: {
    ID: () => 'UNKNOWN_ID',
    Float: () => 0,
    Boolean: () => false,
    String: () => '',
    UUID: () => '63bcbf07-c5e7-4c89-ad07-74030a09f5f6',
  },
  onGuarded() {},
})

describe('plugin', () => {
  it('is applied to the resolver for every field in the schema', async () => {
    const lifecycleCalls: string[] = []
    const beforeCalls: string[] = []
    const afterCalls: string[] = []
    const schema = makeSchema({
      outputs: false,
      types: [buildSchema(EXAMPLE_SDL)],
      plugins: [
        plugin({
          name: 'Lifecycle test',
          onBeforeBuild: () => lifecycleCalls.push('onBeforeBuild'),
          onInstall: () => {
            lifecycleCalls.push('onInstall')
          },
          onCreateFieldResolver({ fieldConfig }) {
            return async (root, args, ctx, info, next) => {
              beforeCalls.push(`${info.parentType.name}:${info.fieldName}`)
              const val = await next(root, args, ctx, info)
              afterCalls.push(`${info.parentType.name}:${info.fieldName} ${val}`)
              return val
            }
          },
          onAfterBuild: (schema) => {
            lifecycleCalls.push('onAfterBuild')
            expect(schema).toBeInstanceOf(GraphQLSchema)
          },
        }),
        nullGuardPlugin,
      ],
      features: {
        abstractTypeStrategies: {
          __typename: true,
        },
      },
    })
    const result = await graphql(
      schema,
      `
        {
          user {
            id
            name
            email
            phone
          }
          posts(filters: { order: DESC }) {
            id
            uuid
            author {
              id
            }
          }
        }
      `
    )
    expect(lifecycleCalls).toMatchSnapshot()
    expect(beforeCalls).toMatchSnapshot()
    expect(afterCalls).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })
  it('throws when the plugin is included in the types but not the plugins array', () => {
    expect.assertions(1)
    try {
      makeSchema({
        outputs: false,
        types: [buildSchema(EXAMPLE_SDL), nullGuardPlugin],
      })
    } catch (e) {
      expect(e.message).toEqual(
        `Nexus plugin NullabilityGuard was seen in the "types" config, but should instead be provided to the "plugins" array.`
      )
    }
  })

  it('does not throw if the plugin is seen in the types as well as the plugins array', () => {
    makeSchema({
      outputs: false,
      types: [buildSchema(EXAMPLE_SDL), nullGuardPlugin],
      plugins: [nullGuardPlugin],
      features: {
        abstractTypeStrategies: {
          __typename: true,
        },
      },
    })
  })

  it('has an onMissingType, which will be called in order when we encounter a missing type', () => {
    const schema = makeSchema({
      outputs: false,
      types: [
        objectType({
          name: 'User',
          definition(t) {
            t.id('id')
          },
        }),
        queryField('users', {
          type: 'UserConnection',
          resolve: () => [],
        }),
      ],
      plugins: [
        plugin({
          name: 'DynamicConnection',
          onMissingType(name, builder) {
            if (name === 'ConnectionInfo') {
              return objectType({
                name,
                definition(t) {
                  t.boolean('hasNextPage')
                  t.boolean('hasPrevPage')
                },
              })
            }
            const exec = /^(.*?)(Connection|Edge|Node)$/.exec(name)
            if (!exec) {
              return null
            }
            const [, typeName, fieldType] = exec
            switch (fieldType) {
              case 'Edge': {
                return objectType({
                  name: name,
                  definition(t) {
                    t.string('cursor')
                    t.field('node', { type: typeName })
                  },
                })
              }
              case 'Connection': {
                return objectType({
                  name: name,
                  definition(t) {
                    t.field('edges', {
                      type: list(`${typeName}Edge`),
                    })
                    t.field('connectionInfo', {
                      type: 'ConnectionInfo',
                    })
                  },
                })
              }
            }
          },
        }),
      ],
    })
    expect(printSchema(lexicographicSortSchema(schema))).toMatchSnapshot()
  })

  it('composes the onCreateFieldResolve fns', async () => {
    const calls: string[] = []
    const testResolve = (name: string) => (): MiddlewareFn => (root, args, ctx, info, next) => {
      calls.push(`Before:${name}`)
      return plugin.completeValue(next(root, args, ctx, info), (val) => {
        calls.push(`After:${name} ${val}`)
        return val + 1
      })
    }
    const schema = makeSchema({
      outputs: false,
      types: [
        queryField('testCompose', {
          type: 'Int',
          resolve: () => {
            calls.push('calls:resolver')
            return 1
          },
        }),
      ],
      plugins: [
        plugin({
          name: 'a',
          onCreateFieldResolver: testResolve('a'),
        }),
        plugin({
          name: 'b',
          onCreateFieldResolver: testResolve('b'),
        }),
        plugin({
          name: 'c',
          onCreateFieldResolver: testResolve('c'),
        }),
      ],
    })
    await graphql(
      schema,
      `
        {
          testCompose
        }
      `
    )
    expect(calls).toMatchSnapshot()
  })

  it('has an onObjectDefinition option, which receives the object metadata', async () => {
    //
    const schema = makeSchema({
      outputs: false,
      features: {
        abstractTypeStrategies: {
          __typename: true,
        },
      },
      types: [
        interfaceType({
          name: 'Node',
          resolveType(n) {
            return n.__typename
          },
          definition(t) {
            t.field('id', {
              type: nonNull('ID'),
              resolve: () => {
                throw new Error('Abstract')
              },
            })
          },
        }),
        objectType({
          name: 'AddsNode',
          // @ts-ignore
          node: 'id',
          definition(t) {
            t.string('name')
          },
        }),
        queryField('getNode', {
          type: 'Node',
          resolve: () => ({ __typename: 'AddsNode', name: 'test', id: 'abc' }),
        }),
      ],
      plugins: [
        plugin({
          name: 'Node',
          onObjectDefinition(t, config) {
            const node = (config as any).node as any
            if (node) {
              t.implements('Node')
              t.field('id', {
                type: nonNull('ID'),
                resolve: (root) => `${config.name}:${root[node]}`,
              })
            }
          },
        }),
      ],
    })
    const result = await graphql(
      schema,
      `
        {
          getNode {
            __typename
            id
            ... on AddsNode {
              name
            }
          }
        }
      `
    )
    expect(result.data?.getNode).toEqual({ __typename: 'AddsNode', id: 'AddsNode:abc', name: 'test' })
  })

  it('has an onAddOutputField / onAddInputField / onAddArg option, which receives the field metadata, and can modify the field', async () => {
    //
    const schema = makeSchema({
      outputs: false,
      features: {
        abstractTypeStrategies: {
          __typename: true,
        },
      },
      types: [
        queryField('ok', {
          type: 'Boolean',
          // @ts-ignore
          listTest: true,
          args: {
            // @ts-ignore
            filter: booleanArg({ listTest: true }),
            input: inputObjectType({
              name: 'SomeType',
              definition(t) {
                // @ts-ignore
                t.boolean('inputField', { listTest: true })
              },
            }),
          },
          resolve: () => [true],
        }),
      ],
      plugins: [
        plugin({
          name: 'Node',
          onAddOutputField(field) {
            // @ts-ignore
            if (field.listTest) {
              field.type = list(field.type)
            }
          },
          onAddInputField(field) {
            // @ts-ignore
            if (field.listTest) {
              field.type = list(field.type)
            }
          },
          onAddArg(arg) {
            // @ts-ignore
            if (arg.listTest) {
              arg.type = list(arg.type)
            }
          },
        }),
      ],
    })
    expect(printSchema(lexicographicSortSchema(schema))).toMatchSnapshot()
  })

  it('has a plugin.completeValue fn which is used to efficiently complete a value which is possibly a promise', async () => {
    const calls: string[] = []
    const testResolve = (name: string) => (): MiddlewareFn => async (root, args, ctx, info, next) => {
      calls.push(`Before:${name}`)
      return plugin.completeValue(next(root, args, ctx, info), (val) => {
        calls.push(`After:${name} ${val}`)
        return val + 1
      })
    }
    const schema = makeSchema({
      outputs: false,
      types: [
        queryField('testCompose', {
          type: 'Int',
          resolve: async () => {
            calls.push('calls:resolver')
            return 1
          },
        }),
      ],
      plugins: [
        plugin({
          name: 'a',
          onCreateFieldResolver: testResolve('a'),
        }),
        plugin({
          name: 'b',
          onCreateFieldResolver: testResolve('b'),
        }),
        plugin({
          name: 'c',
          onCreateFieldResolver: testResolve('c'),
        }),
      ],
    })
    await graphql(
      schema,
      `
        {
          testCompose
        }
      `
    )
    expect(calls).toMatchSnapshot()
  })
})
