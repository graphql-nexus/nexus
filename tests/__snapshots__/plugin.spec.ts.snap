// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin composes the onCreateFieldResolve fns 1`] = `
Array [
  "Before:a",
  "Before:b",
  "Before:c",
  "calls:resolver",
  "After:c 1",
  "After:b 2",
  "After:a 3",
]
`;

exports[`plugin has a plugin.completeValue fn which is used to efficiently complete a value which is possibly a promise 1`] = `
Array [
  "Before:a",
  "Before:b",
  "Before:c",
  "calls:resolver",
  "After:c 1",
  "After:b 2",
  "After:a 3",
]
`;

exports[`plugin has an onAddOutputField / onAddInputField / onAddArg option, which receives the field metadata, and can modify the field 1`] = `
"type Query {
  ok(filter: [Boolean], input: SomeType): [Boolean]
}

input SomeType {
  inputField: [Boolean]
}
"
`;

exports[`plugin has an onMissingType, which will be called in order when we encounter a missing type 1`] = `
"type ConnectionInfo {
  hasNextPage: Boolean
  hasPrevPage: Boolean
}

type Query {
  users: UserConnection
}

type User {
  id: ID
}

type UserConnection {
  connectionInfo: ConnectionInfo
  edges: [UserEdge]
}

type UserEdge {
  cursor: String
  node: User
}
"
`;

exports[`plugin is applied to the resolver for every field in the schema 1`] = `
Array [
  "onInstall",
  "onBeforeBuild",
  "onAfterBuild",
]
`;

exports[`plugin is applied to the resolver for every field in the schema 2`] = `
Array [
  "Query:user",
  "Query:posts",
  "User:id",
  "User:name",
  "User:email",
  "User:phone",
]
`;

exports[`plugin is applied to the resolver for every field in the schema 3`] = `
Array [
  "Query:user [object Object]",
  "Query:posts ",
  "User:id UNKNOWN_ID",
  "User:name ",
  "User:email ",
  "User:phone undefined",
]
`;

exports[`plugin is applied to the resolver for every field in the schema 4`] = `
Object {
  "data": Object {
    "posts": Array [],
    "user": Object {
      "email": "",
      "id": "UNKNOWN_ID",
      "name": "",
      "phone": null,
    },
  },
}
`;
